### ISSUE TITLE
-s PTHREAD_POOL_SIZE = ... 

### ISSUE THREAD

**Issue by yangfangfang1204** [emscripten-core/emscripten#14407]
I use threads in video decoding, in a practical application scenario, I'm not sure how many decoding objects will be created, so I don't know how many threads will be required?
And what the parameters `-s PTHREAD_POOL_SIZE = ... ` should be set to?100?
However, if fewer than 100 subThreads are used, this will increase the memory consumption of the browser? Such as If i need two threads ,i set `-s PTHREAD_POOL_SIZE = 2` , the chrome used memory is 200MB; and if i set `-s PTHREAD_POOL_SIZE = 20` , the chrome used memory is 400MB . 
So do you have other method to solve the problem？

**Comment by yangfangfang1204** [comment]
And if pool_size Less than the number of threads actually needed, there is an warning 
`Tried to spawn a new thread, but the thread pool is exhausted.
This might result in a deadlock unless some threads eventually exit or the code explicitly breaks out to the event loop.
If you want to increase the pool size, use setting `-s PTHREAD_POOL_SIZE=...`.
If you want to throw an explicit error instead of the risk of deadlocking in those cases, use setting `-s PTHREAD_POOL_SIZE_STRICT=2`.`

**Comment by sbc100** [comment]
You you want to avoid the problem completely you can set `PTHREAD_POOL_SIZE_STRICT=0` and new threads will be created on-demand.

The limitation with creating on-demend threads is that you interacting with the thread from the main browser thread requires to you to yield to the event loop.   If you return to even loop frequently, or if you use PROZY_TO_PTHREAD, or if you create your threads from the secondary thread, then you should be fine.

In short, if you avoid blocking on threads (using pthread_join or mutex_lock) from the main browser thread, then you don't need the thread pool at all.

**Comment by yangfangfang1204** [comment]
If the thread is not a detached thread, how do you exit the thread? Use pthread_exit(0) on the subThread or pthread_join on the main browser thread ?

**Comment by sbc100** [comment]
I thread can exit by returning from its entry function or by calling pthread_exit.  This is true for both detached threads and non-detached thread.    For non-detached threads there will some kind amount of state that is maintained after thread exit up until the point that it is joined by some other thread (no necessarily the creating thread or the main thread).

**Comment by RReverser** [comment]
Realistically, you don't want to create more threads than there are actual CPU cores. It's not as just about memory consumption as about the startup cost and synchronization as well.

So I'd recommend to set `-s PTHREAD_POOL_SIZE=navigator.hardwareConcurrency` which will create as many threads as there are cores, and then use [`std::thread::hardware_concurrency`](https://www.cplusplus.com/reference/thread/thread/hardware_concurrency/) on the C++ side to schedule tasks.

**Comment by yangfangfang1204** [comment]
If the number of threads I need is 10, and the actual CPU kernel only has 6. If `-s PTHREAD_POOL_SIZE=navigator.hardwareConcurrency`  is added at compile time.
At the runtime ,  a warning will appear `Tried to spawn a new thread, but the thread pool is exhausted. This might result in a deadlock unless some threads eventually exit or the code explicitly breaks out to the event loop. If you want to increase the pool size, use setting -s PTHREAD_POOL_SIZE=... If you want to throw an explicit error instead of the risk of deadlocking in those cases, use setting -s PTHREAD_POOL_SIZE_STRICT=2` ;
But  10 threads will be created, as you can see on the Souces page with 10 *.worker.js ?

**Comment by RReverser** [comment]
> If the number of threads I need is 10

Right, but the question is why do you need 10 threads if CPU has only 6? Normally this kind of work is instead scheduled via a [work-stealing queue](https://en.wikipedia.org/wiki/Work_stealing) so that you create only 6 threads - up to CPU's limit - and then each thread takes a task from the queue as soon as it becomes available.

This way you get the best balance between utilising multithreading and avoiding unnecessary overhead from creating too many threads.

**Comment by yangfangfang1204** [comment]
I'm decoding high-resolution video, which takes five threads to stream.Ten threads are required if the live environment has two streams.
My question is, why did I create 10 threads successfully on a computer with only 6 actual cores on the CPU?

**Comment by RReverser** [comment]
> My question is, why did I create 10 threads successfully on a computer with only 6 actual cores on the CPU?

If you're asking about native code - it's because you can create as many native threads as you want, it's just that you're forcing OS to do all the scheduling and context-switching between them, which is much more expensive than doing a local queue.

**Comment by RReverser** [comment]
Anyway, if you know you'll always have 10 threads, you can use `-s PTHREAD_POOL_SIZE=10`. I'm just saying it's best not to create threads proportionally to the number of inputs which it sounds like what you were trying to do:

> I use threads in video decoding, in a practical application scenario, I'm not sure how many decoding objects will be created, so I don't know how many threads will be required?
And what the parameters -s PTHREAD_POOL_SIZE = ... should be set to?100?

It's an easy way to DDOS yourself. Instead, it's best to use a queue with work-stealing technique.

**Comment by goldwaving** [comment]
I've noticed that setting PTHREAD_POOL_SIZE has a significant impact on initial startup time.  If your main.js file is large, that file will be downloaded multiple times depending on the pool size you set (preloaded .data files may be downloaded multiple times as well).  In my current project, using PTHREAD_POOL_SIZE=10 caused startup time to take over 10 times longer.  Caching and service workers help on the second run, but there is still a performance reduction on startup.  Using PTHREAD_POOL_SIZE_STRICT=0 was the best option in my project.

**Comment by RReverser** [comment]
> Caching and service workers help on the second run

That's odd. It's true that setting high value for pool size is not recommended - that's why I said above it's best to set it to number of CPU cores and not depending on the input size - but caching should kick in even on the first run. After browser downloaded `main.js` for the main thread, it can reuse it for Workers too.

Are you sure you weren't testing with DevTools open and "Disable cache" toggle checked? (asking just in case, because it's a fairly common oversight) Alternatively, are caching headers correctly set by your server you used for testing?

**Comment by sbc100** [comment]
The best way to avoid the initial cost of the pool is not to use one at all.    Its not always possible, but you can avoid the pool completely if you structure you app in such a way that it doesn't start threads from the main browser thread and then immediately try to communicate with them. 

One way to avoid needing a pool at all is to use PROXY_TO_PTHREAD to move your main function itself to a thread.   I think that that removes the need for the pool completely.   Of course the overall cost of starting the threads is the same whether you front load it by using a pool or do it on demand as you create you threads.

**Comment by goldwaving** [comment]
@RReverser Part of the problem may have been a server issue, so that could account for the multiple requests.  If I had more time, I would run some more tests.  I also had to use ignoreVary in the service worker to get caching to work at all.  The hosting provider we are using has very little support for progressive web apps.

@sbc100 I'm using a thread for the main program, so I was able to avoid using the pool.

**Comment by sbc100** [comment]
Ok great!  Sorry I misunderstood.   So is the issue is solved now?

**Comment by goldwaving** [comment]
Yes.  I now believe my startup issue was server related.  Even better now that I don't need the pool.  Thanks!

**Comment by RReverser** [comment]
@sbc100 They weren't the reporter of this issue though.

**Comment by sbc100** [comment]
@yangfangfang1204 as you satisfied with the answers in this thread?

**Comment by yangfangfang1204** [comment]
i use `-s PTHREAD_POOL_SIZE_STRICT =0` slove the problem.

**Comment by sunlong0724** [comment]
It works!

**Comment by sunlong0724** [comment]
> I use threads in video decoding, in a practical application scenario, I'm not sure how many decoding objects will be created, so I don't know how many threads will be required? And what the parameters `-s PTHREAD_POOL_SIZE = ... ` should be set to?100? However, if fewer than 100 subThreads are used, this will increase the memory consumption of the browser? Such as If i need two threads ,i set `-s PTHREAD_POOL_SIZE = 2` , the chrome used memory is 200MB; and if i set `-s PTHREAD_POOL_SIZE = 20` , the chrome used memory is 400MB . So do you have other method to solve the problem？

@yangfangfang1204 What compile flags are used when you complie wasm by emsdk？
